lib/puma/request.rb

```diff
@@ -160,34 +163,40 @@ def prepare_response(status, headers, res_body, requests, client)
 
       resp_info = str_headers(env, status, headers, res_body, io_buffer, force_keep_alive)
 
+      close_body = false
+
       # below converts app_body into body, dependent on app_body's characteristics, and
       # resp_info[:content_length] will be set if it can be determined
       if !resp_info[:content_length] && !resp_info[:transfer_encoding] && status != 204
-        if res_body.respond_to?(:to_ary)
-          length = 0
-          if array_body = res_body.to_ary
-            body = array_body.map { |part| length += part.bytesize; part }
-          elsif res_body.is_a?(::File) && res_body.respond_to?(:size)
-            length = res_body.size
-          elsif res_body.respond_to?(:each)
-            body = []
-            res_body.each { |part| length += part.bytesize; body << part }
-          end
-          resp_info[:content_length] = length
+        if res_body.respond_to?(:to_ary) && (array_body = res_body.to_ary) && array_body.is_a?(Array)
```

Comment:
```suggestion
        if res_body.respond_to?(:to_ary) && (array_body = res_body.to_ary)
```

Reply:
I'm going to test this.  I recall that one of the Rails wrapped responses/bodies returned `nil` when `to_ary` was called.  Not.sure.

Kicking myself for not adding a comment about it.  Give me a bit.

Reply:
Ahh I see... I think if `to_ary` returns nil, you can assume there is no body. This is out of alignment with the Rack spec in any case.

Reply:
Fixerd.  Pushed back into 'Updates for response body processing'